src/typ.rs

```diff
@@ -2278,6 +2288,94 @@ impl<'a> Typer<'a> {
         })
     }
 
+    fn infer_record_update(
```

Comment:
There's a lot going on in this function, and I'd still like to get some comments describing what I'm doing. It may also be worthwhile to flatten things a little bit, if you have any ideas on doing so nicely :)